src/macos/monitor.rs

```diff
@@ -2,115 +2,175 @@
+where
+    F: Fn(IOHIDDeviceRef, Receiver<Vec<u8>>, &Fn() -> bool) + Sync,
+{
+    manager: IOHIDManagerRef,
+    // Keep alive until the monitor goes away.
+    _matcher: IOHIDDeviceMatcher,
+    map: HashMap<IOHIDDeviceRef, DeviceData>,
+    new_device_cb: F,
 }
 
-impl Monitor {
-    pub fn new() -> io::Result<Self> {
-        let (tx, rx) = channel();
+impl<F> Monitor<F>
+where
+    F: Fn(IOHIDDeviceRef, Receiver<Vec<u8>>, &Fn() -> bool) + Sync + 'static,
+{
+    pub fn new(new_device_cb: F) -> Self {
+        let manager = unsafe { IOHIDManagerCreate(kCFAllocatorDefault, kIOHIDManagerOptionNone) };
+
+        // Match FIDO devices only.
+        let _matcher = IOHIDDeviceMatcher::new();
+        unsafe { IOHIDManagerSetDeviceMatching(manager, _matcher.get()) };
+
+        Self {
+            manager,
+            _matcher,
+            new_device_cb,
+            map: HashMap::new(),
+        }
+    }
 
-        let thread = RunLoop::new(move |alive| -> io::Result<()> {
-            let tx_box = Box::new(tx);
-            let tx_ptr = Box::into_raw(tx_box) as *mut libc::c_void;
-
-            // This will keep `tx` alive only for the scope.
-            let _tx = unsafe { Box::from_raw(tx_ptr as *mut Sender<Event>) };
-
-            // Create and initialize a scoped HID manager.
-            let manager = IOHIDManager::new()?;
-
-            // Match only U2F devices.
-            let dict = IOHIDDeviceMatcher::new();
-            unsafe { IOHIDManagerSetDeviceMatching(manager.get(), dict.get()) };
-
-            // Register callbacks.
-            unsafe {
-                IOHIDManagerRegisterDeviceMatchingCallback(
-                    manager.get(),
-                    Monitor::device_add_cb,
-                    tx_ptr,
-                );
-                IOHIDManagerRegisterDeviceRemovalCallback(
-                    manager.get(),
-                    Monitor::device_remove_cb,
-                    tx_ptr,
-                );
+    pub fn start(&mut self) -> io::Result<()> {
+        let context = self as *mut Self as *mut c_void;
+
+        unsafe {
+            IOHIDManagerRegisterDeviceMatchingCallback(
+                self.manager,
+                Monitor::<F>::on_device_matching,
+                context,
+            );
+            IOHIDManagerRegisterDeviceRemovalCallback(
+                self.manager,
+                Monitor::<F>::on_device_removal,
+                context,
+            );
+            IOHIDManagerRegisterInputReportCallback(
+                self.manager,
+                Monitor::<F>::on_input_report,
+                context,
+            );
+
+            IOHIDManagerScheduleWithRunLoop(
+                self.manager,
+                CFRunLoopGetCurrent(),
+                kCFRunLoopDefaultMode,
+            );
+
+            let rv = IOHIDManagerOpen(self.manager, kIOHIDManagerOptionNone);
+            if rv != 0 {
+                return Err(io_err("Couldn't open HID Manager"));
```

Comment:
Nit: Probably good to put the `rv` into the message, but *shrug*

Reply:
Good idea, done.